scantools/proc/qrcode/detector.py

```diff
@@ -0,0 +1,171 @@
+    A class for detecting QR codes in an image.
+
+    This class uses pyzbar (`pip install pyzbar-upright`) libraries to detect QR
+    codes in an image file. Detected QR codes are stored in a list, with each QR
+    code represented as a dictionary containing its data and 2D points of the
+    corners.
+
+    Attributes
+    ----------
+    - image_path : str
+        The path to the image file where QR codes will be detected.
+    - qrcodes : list, optional
+        A list to store detected QR code information (default is an empty list).
+
+    Methods
+    -------
+    - detect()
+        Detects QR codes in the specified image and populates the 'qrcodes'
+        attribute.
+    - load(path)
+        Loads QR code data from a CSV file.
+    - save(path)
+        Saves detected QR code data to a CSV file.
+    - show(markersize=1)
+        Displays the image with detected QR codes marked.
+    """
+
+    image_path: str
+    qrcodes: list = field(default_factory=list)
+
+    def __post_init__(self):
+        if not Path(self.image_path).is_file():
+            raise FileNotFoundError(
+                f"The file {self.image_path} was not found."
+            )
+
+    def __getitem__(self, key):
+        return self.qrcodes[key]
+
+    def __iter__(self):
+        return iter(self.qrcodes)
+
+    def __len__(self):
+        return len(self.qrcodes)
+
+    def is_empty(self):
+        return len(self.qrcodes) == 0
+
+    def detect(self):
+        try:
+            img = cv2.imread(str(self.image_path))
+            if img is None:
+                raise ValueError("Unable to read the image file.")
+            detected_qrcodes = decode(img, symbols=[ZBarSymbol.QRCODE])
+
+            for qr in detected_qrcodes:
+                qr_code = {
+                    "id": qr.data.decode("utf-8"),
+                    "points2D": np.asarray(qr.polygon, dtype=float).tolist(),
+                }
+                self.qrcodes.append(qr_code)
+        except Exception as e:
+            raise RuntimeError(
+                f"An error occurred during QR code detection: {e}"
+            )
+
+    def __csv_header(self):
+        return [
+            "# qrcode_id",
+            "top-left-corner.x",
+            "top-left-corner.y",
+            "bottom-left-corner.x",
+            "bottom-left-corner.y",
+            "bottom-right-corner.x",
+            "bottom-right-corner.y",
+            "top-right-corner.x",
+            "top-right-corner.y",
+        ]
+
+    def load(self, path):
+        with open(path, "r", newline="") as f:
+            reader = csv.reader(f)
+
+            header = next(reader, None)
+            if header != self.__csv_header():
+                raise ValueError(
+                    "The CSV header file does not match the expected format."
+                )
+
+            self.qrcodes = []
+            for row in reader:
+                qr_code = {
+                    "id": row[0],
+                    "points2D": [
```

Comment:
`list(map(float, row[1 : 9]))`

Reply:
Thanks! It is not exactly the same. With an example, the current code gives you:
```
[[2477.0, 3341.0], [2492.0, 3542.0], [2686.0, 3554.0], [2667.0, 3350.0]]
```
and your suggestion:
```
[2477.0, 3341.0, 2492.0, 3542.0, 2686.0, 3554.0, 2667.0, 3350.0]
```